src/rust/src/padding.rs

```diff
@@ -110,6 +110,62 @@ impl PKCS7PaddingContext {
     }
 }
 
+#[pyo3::pyclass]
+pub(crate) struct _ANSIX923PaddingContext {
+    block_size: usize,
+    buffer: Option<Vec<u8>>,
+}
+
+#[pyo3::pymethods]
+impl _ANSIX923PaddingContext {
+    #[new]
+    pub(crate) fn new(block_size: usize) -> _ANSIX923PaddingContext {
+        _ANSIX923PaddingContext {
+            block_size,
+            buffer: Some(Vec::new()),
+        }
+    }
+
+    pub(crate) fn update<'p>(
+        &mut self,
+        buf: CffiBuf<'p>,
+        py: pyo3::Python<'p>,
+    ) -> CryptographyResult<pyo3::Bound<'p, pyo3::types::PyBytes>> {
+        match self.buffer.as_mut() {
+            Some(v) => {
+                v.extend_from_slice(buf.as_bytes());
+                let finished_blocks = v.len() / (self.block_size / 8);
+                let result_size = finished_blocks * (self.block_size / 8);
+                let result = v.drain(..result_size);
+                Ok(pyo3::types::PyBytes::new(py, result.as_slice()))
+            }
+            None => Err(exceptions::already_finalized_error()),
+        }
+    }
+
+    pub(crate) fn finalize<'p>(
+        &mut self,
+        py: pyo3::Python<'p>,
+    ) -> CryptographyResult<pyo3::Bound<'p, pyo3::types::PyBytes>> {
+        match self.buffer.take().as_mut() {
+            Some(v) => {
+                let pad_size = (self.block_size / 8).saturating_sub(v.len());
+                let mut count = 0;
+                v.extend(std::iter::from_fn(move || {
+                    count += 1;
+                    match count.cmp(&pad_size) {
+                        std::cmp::Ordering::Less => Some(0),
+                        std::cmp::Ordering::Equal => Some(pad_size as u8),
```

Comment:
I think this truncating cast to u8 is fine by construction but I can't see how to prove that.